### ISSUE TITLE
Rule Change: [indent] Basic support for smart tabs

### ISSUE THREAD

**Issue by DMartens** [eslint/eslint#16281]
### What rule do you want to change?

indent

### What change to do you want to make?

Generate more warnings

### How do you think the change should be implemented?

A new option

### Example code

```js
// Additional indentation
\t const shouldHaveNoIndentation = true;
^^^

// Missing indentation
function a() {
\tfunction b() {
\t return "Should have an indentation of two tabs"
^^^
\t}
}

// Statements should not be aligned with spaces
function fn() {
\t const shouldNotHaveASpaceBefore = true;
}

// Not aligned to an expression on a prior line
function fn(array) {
\tconst mapped = array.map(() => {
\t }); // Notice the extra space
  ^
}
```

### What does the rule currently do for this code?

Does not report in any of the mentioned cases.

### What will the rule do after it's changed?

Currently any line containing mixed tabs and spaces is not checked to not interfere with `no-mixed-spaces-and-tabs` (see #7248).
But if `no-mixed-spaces-and-tabs` allows smart tabs (tabs for indentation, spaces for alignment),  the mentioned errors may go unnoticed.
This is why I propose to add a new option `checkSmartTabs`.
When the option `checkSmartTabs` is set, the following three checks would be added to the rule if the line indentation uses smart tabs:
First the expected indentation (which consists of only tabs) is matched with only the tabs of indentation and reports if the amount of tabs is different. This handles the first two cases and would be fixable.
The second check would be that smart tabs can only be used for expressions and not for statements. The fix for this could simply remove the spaces from the indentation.
The last check would be if there is a token at the same column but on a prior line inside the node to make sure that the node is aligned.
This would still cause false negatives if the expression is not aligned with the expected expression but this would require options for each node type to know to what to align to and this could also affect users using only spaces. So I would not include alignment in the current proposal.
This last check could also be in another rule as no indentation information is required.

### Participation

- [X] I am willing to submit a pull request to implement this change.

### Additional comments

I added the checks as a proof of concept but another problem arises from the fact that the user has no control when to add indentation, only to ignore the node completely.
For example adding an indentation level for initial values of variables would mess with smart tabs as the alignment would differ (if the space difference is not exactly a tab width).
This would require an additional option to control which of the checked node types can add an indentation level.

**Comment by mdjermanovic** [comment]
> Currently any line containing mixed tabs and spaces is not checked to not interfere with `no-mixed-spaces-and-tabs` (see #7248).

As I understand it, this behavior is intentional (https://github.com/eslint/eslint/pull/7266) and the rule works as intended (i.e., this is not a bug). Per our rules policies, [stylistic rules are frozen](https://github.com/eslint/eslint#stylistic-rule-updates) so we're not adding new options to the `indent` rule. You can always copy the existing rule into a custom rule and make it behave the way you want.
